### PR TITLE
Support full drop/delete in storage and indexing backends

### LINKED ISSUE #462: Support graph/index deletion in core factory API
This issue is based on discussions in #392 and in https://groups.google.com/forum/#!topic/janusgraph-dev/hYFY-_nshyg (thanks @dpitera). To summarize the core graph factory API should support graph/index deletion with the expected result that keyspaces/tables/indices/etc. are dropped/deleted.

Currently storage and indexing backends implement deletion through [Backend#clearStorage()](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/StoreManager.java#L48) and [IndexProvider#clearStorage()](https://github.com/JanusGraph/janusgraph/blob/7f924fade7cc36912a9ad9f2883011c7c511d72c/janusgraph-core/src/main/java/org/janusgraph/diskstorage/indexing/IndexProvider.java#L131), respectively, which is exposed at a higher level through [JanusGraphCleanup#clear()](https://github.com/JanusGraph/janusgraph/blob/bfa3c970555932d4983155168322f72297aaee85/janusgraph-core/src/main/java/org/janusgraph/core/util/JanusGraphCleanup.java#L42). But existing delete implementations are not generally consistent with the expected behavior as it appears they were written to support the testing rather than production use case (`JanusGraphCleanup#clear()` documentation indicates that the method "is intended only for development and testing use"). For example the Cassandra implementations (CQL, Thrift, etc.) truncate instead of drop keyspaces, the HBase implementation deletes rows using an implementation that does not scale, and the Solr implementation does not appear to delete collections as would be expected. Also tests are missing across all storage/indexing backends to validate expected delete behavior.

The goal in addressing this issue should be to support dropping keyspaces/etc. while still maintaining the existing clearStorage implementations for testing. The latter is critical because it's been found that dropping keyspaces/tables during testing has a significant impact on performance (see #384 which showed truncation, among other updates, decreased runtime for the TinkerPop test suite in the CQL module from 15 hours to 3 hours).

### PR DESCRIPTION
**Pull request by sjudeng**
Fixes #462

### CODE REVIEW

**Comment by sjudeng** [comment]
The build failure is due to Codecov. I thought the status message/checks were disabled in [codecov.yml](https://github.com/JanusGraph/janusgraph/blob/master/codecov.yml). Will look into that separately from this PR.

File: janusgraph-cql/src/test/java/org/janusgraph/diskstorage/cql/CachingCQLStoreManager.java [hunk 1]
```diff
@@ -36,6 +40,13 @@ Session initializeSession(final String keyspaceName) {
 
     @Override
     public void close() {
+        if (this.storageConfig.get(DROP_ON_CLEAR)) {
+            final Iterator<Map.Entry<String, Session>> iterator = sessions.entrySet().iterator();
+            while (iterator.hasNext()) {
```
Comment (davidclement90) [inline janusgraph-cql/src/test/java/org/janusgraph/diskstorage/cql/CachingCQLStoreManager.java]: You can use stream to make it more compact.

**Comment by sjudeng** [comment]
@davidclement90 Did you receive (and approve) the invitation mentioned in https://github.com/JanusGraph/janusgraph/pull/490#issuecomment-327361360? @mbrukman Anything else needed to get David write access?

**Comment by mbrukman** [comment]
@sjudeng — it appears that @davidclement90 hasn't yet accepted his invitation; once that's done, he automatically gets write access and his checkmarks get a shiny coat of green paint.

**Comment by sjudeng** [comment]
@davidclement90 Please do approve that invitation since otherwise unfortunately your review does not count for the GitHub merge check.

**Comment by mbrukman** [comment]
@sjudeng – per @davidclement90's [comment on another PR](https://github.com/JanusGraph/janusgraph/pull/490#issuecomment-329805917), the GitHub committer invitation ended up in his spam folder, so he did not notice it earlier. Now that he's accepted it, his approval checkmarks are green everywhere, and he should have write / merge access to this repo.

**Comment by davidclement90** [comment]
The mail was in my spam...
I just accept it.

**Pull request merged**